internal/unionstore/art/art_iterator.go

```diff
@@ -156,140 +156,161 @@ func (it *Iterator) init(lowerBound, upperBound []byte) {
+type baseIter struct {
+	allocator *artAllocator
+	// the baseIter iterate the ART tree in DFS order, the idxes and nodes are the current visiting stack.
+	idxes []int
+	nodes []artNode
+}
+
+// seekToFirst seeks the boundary of the tree, sequential or reverse.
+func (it *baseIter) seekToFirst(root artNode, reverse bool) {
+	// if the seek key is empty, it means -inf or +inf, return root node directly.
+	it.nodes = []artNode{root}
+	if reverse {
+		it.idxes = []int{node256cap}
+	} else {
+		it.idxes = []int{inplaceIndex}
+	}
 }
 
-// seek the first node and index that >= key, return the indexes and nodes of the lookup path
-// nodes[0] is the root node
-func (it *Iterator) seek(key artKey) ([]int, []artNode) {
-	curr := it.tree.root
+// seek the first node and index that >= key, nodes[0] is the root node
+func (it *baseIter) seek(root artNode, key artKey) {
+	curr := root
 	depth := uint32(0)
-	idxes := make([]int, 0, 8)
-	nodes := make([]artNode, 0, 8)
+	it.idxes = make([]int, 0, 8)
+	it.nodes = make([]artNode, 0, 8)
 	if len(key) == 0 {
-		// if the seek key is empty, it means -inf or +inf, return root node directly.
-		nodes = append(nodes, curr)
-		if it.reverse {
-			idxes = append(idxes, node256cap)
-		} else {
-			idxes = append(idxes, inplaceIndex)
-		}
-		return idxes, nodes
+		panic("seek with empty key is not allowed")
 	}
 	var node *nodeBase
 	for {
 		if curr.isLeaf() {
 			if key.valid(int(depth)) {
-				lf := curr.asLeaf(&it.tree.allocator)
+				lf := curr.asLeaf(it.allocator)
 				if bytes.Compare(key, lf.GetKey()) > 0 {
 					// the seek key is not exist, and it's longer and larger than the current leaf's key.
 					// e.g. key: [1, 1, 1], leaf: [1, 1].
-					idxes[len(idxes)-1]++
+					it.idxes[len(it.idxes)-1]++
 				}
 			}
-			break
+			return
 		}
 
-		node = curr.asNode(&it.tree.allocator)
+		node = curr.asNode(it.allocator)
 		if node.prefixLen > 0 {
-			mismatchIdx := node.matchDeep(&it.tree.allocator, &curr, key, depth)
+			mismatchIdx := node.matchDeep(it.allocator, &curr, key, depth)
 			if mismatchIdx < node.prefixLen {
-				// no leaf node is match with the seek key
-				leafNode := minimum(&it.tree.allocator, curr)
-				leafKey := leafNode.asLeaf(&it.tree.allocator).GetKey()
-				if mismatchIdx+depth == uint32(len(key)) || key[depth+mismatchIdx] < leafKey[depth+mismatchIdx] {
+				// no leaf node is match with the seek key, we need to check whether the seek key is less than the prefix.
+				// If the seek key is less than the prefix, all the children are located on the right side of the seek key,
+				// unless the children are located on the left side of the seek key.
+				var prefix []byte
+				if mismatchIdx < maxPrefixLen {
+					prefix = node.prefix[:]
+				} else {
+					leafNode := minimum(it.allocator, curr)
+					prefix = leafNode.asLeaf(it.allocator).getKeyDepth(depth)
+				}
+				if mismatchIdx+depth == uint32(len(key)) || key[depth+mismatchIdx] < prefix[mismatchIdx] {
```

Comment:
What does it mean if the `mismatchIdx+depth == uint32(len(key))` hold? 
As well as  `mismatchIdx+depth > uint32(len(key))`  or  `mismatchIdx+depth < uint32(len(key))`.

Reply:
If the `mismatchIdx+depth == uint32(len(key))` holds, it means the seek key is the prefix of the any leaf key in this node, and the leaf key is longer and larger than the seek key.